### PR TITLE
Includes file, line and column number in Scanner

### PR DESCRIPTION
**Pull request by andrewemeryanz**
This change also modifies the Scanner internally to maintain the slice
range for derived scanner instances, not the sliced string subset.

### CODE REVIEW

File: parser/scanner.go [hunk 1]
```diff
@@ -7,75 +7,98 @@ import (
...
+	sliceStart  int    // the start of the slice visible to the scanner
+	sliceLength int    // the length of the slice visible to the scanner
 }
 
-func NewScanner(src string) *Scanner {
-	return &Scanner{src: src, slice: src, offset: 0}
+type source interface {
+	length() int                // the length of the entire source string
+	slice(i, length int) string // the string of the given slice
+	filename() *string          // the name of the file from which the source is derived (if any)
```
Comment (marcelocantos) [inline parser/scanner.go]: minor: empty string would suffice to convey "no file".

**Review by marcelocantos — DISMISSED** [review]

File: parser/scanner.go (hunk 1 above)
Comment (andrewemeryanz) [reply parser/scanner.go]: Agreed. Modified.

File: parser/scanner.go [hunk 2]
```diff
@@ -7,75 +7,99 @@ import (
...
+	return s.slice()
+}
+
+func (s Scanner) Format(state fmt.State, c rune) {
 	if c == 'q' {
-		fmt.Fprintf(state, "%q", r.String())
+		_, _ = fmt.Fprintf(state, "%q", s.slice())
 	} else {
-		state.Write([]byte(r.String()))
+		_, _ = state.Write([]byte(s.slice()))
```
Comment (nofun97) [inline parser/scanner.go:56]: nit: Does this need the `_, _ = ...`?

File: parser/scanner.go [hunk 3]
```diff
@@ -7,75 +7,99 @@ import (
...
+	return lineColumn(s.src.slice(0, s.sliceStart), s.sliceStart)
+}
+
+// The slice that is visible to the scanner
+func (s Scanner) slice() string {
+	return s.src.slice(s.sliceStart, s.sliceLength)
+}
+
+func (s Scanner) Slice(a, b int) *Scanner {
+	return &Scanner{s.src, s.sliceStart + a, b - a}
```
Comment (nofun97) [inline parser/scanner.go:85]: shouldn't this be `b - a + 1`

**Review by nofun97 — COMMENTED** [review]
Other than these, LGTM

File: parser/scanner.go (hunk 3 above)
Comment (andrewemeryanz) [reply parser/scanner.go:85]: No. A slice from 0 to 0 should have a length of 0, not 1.

File: parser/scanner.go (hunk 2 above)
Comment (andrewemeryanz) [reply parser/scanner.go:56]: The previous intent was to ignore the return values (which includes an error). This just makes it explicit.

**Review by nofun97 — APPROVED** [review]
LGTM

**Pull request merged**